include/pybind11/pybind11.h

```diff
@@ -1505,14 +1505,22 @@ class class_ : public detail::generic_type {
     template <typename T>
     static void init_holder(detail::instance *inst, detail::value_and_holder &v_h,
             const holder_type * /* unused */, const std::enable_shared_from_this<T> * /* dummy */) {
+#if defined(PYBIND11_CPP17)
+        if (!v_h.value_ptr<type>()->weak_from_this().expired()) {
```

Comment:
This looks good, but do we need to also guard it with the feature test macro? MSVC, being sloppy with feature test macros until ~recently, would need a version check instead.

- For MSVC, it's supported since 19.12.
- For normal compilers, that's `__cpp_lib_enable_shared_from_this >= 201603`.

Reply:
To not clutter code too much I usually wait until there's an actual problem. Our very comprehensive CI is green. How likely is it that those macros are actually useful?

Reply:
Well, @bstaletic's solution is the correct thing to do, if we want to support all compiler ranges we say we support.

Reply:
We also already have this kind of things. It's not really more clutter: https://github.com/pybind/pybind11/blob/587d5f840a234fc11475952c81ee68525875239f/include/pybind11/pybind11.h#L1197

Reply:
Here's a simple use case that would break. Start by using Visual Studio 15.3, which comes with MSVC 19.10. Pass `/std:c++latest` to the compiler and watch it error on `weak_from_this()` not being a thing.

Reply:
The problem for us is that we don't have a practical way of testing the extra `#if defined`. It would be a shot in the blue. If you have access to "Visual Studio 15.3, which comes with MSVC 19.10", could you please help out pushing a commit with the required change to this PR?

Reply:
I'm assuming @bstaletic found some official reference saying when MSVC added this feature?

These kinds of things are always tricky anyway. But I don't see how an informed, reasoned-about decision is worse/more of a stab in the dark than the current blunt `PYBIND11_CPP17`? Even if we're still wrong, we won't be more wrong than saying "C++17". And that's even more the case for when we can use the official `__cpp_lib_enable_shared_from_this`.

And pybind11 is already doing this in various other places, anyway. So if both these pass tests, let's pick the one we can reason about to be more correct than the other.
Unless there's a case where @bstaletic's solution can lead to dangerous/unpredictable situtations, ofc!

Reply:
I mean, everyone can double check my sources:

1. [MSVC 19.12](https://devblogs.microsoft.com/cppblog/c17-progress-in-vs-2017-15-5-and-15-6/)
2. [MSVC 19.11](https://blogs.msdn.microsoft.com/vcblog/2017/08/11/c17-features-and-stl-fixes-in-vs-2017-15-3/)
3. [MSVC 19.10](https://devblogs.microsoft.com/cppblog/vs-2015-update-2s-stl-is-c17-so-far-feature-complete/)
4. [MSVC versioning](https://en.wikipedia.org/wiki/MSVC#Internal_version_numbering)
5. [`__cpp_lib_enable_shared_from_this` in the standard](https://eel.is/c++draft/version.syn#lib:__cpp_lib_enable_shared_from_this)
6. [Relevant committee paper](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0033r1.html)
7. [cppreference on feature test macros](https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros#C.2B.2B17)

And this is the error we can currently expect: https://godbolt.org/z/KKdTnE
Yes, that is with MSVC 19.10, not 19.11, but we can expect the same error on both.

Reply:
@elkhrt You missed the `&& (!defined(_MSC) || _MSC >=1912)` part.

Also, since this is a standard library feature test macro, since C++20, it has been moved to the `<version>` header. So we also need something like:

```
#if defined(__has_include)
#  if __has_include(<version>)
#    include <version>
# endif
#endif
```

Probably somewhere in `detail/common.h`.

And no, `defined(__has_include) && __has_include(<version>)` doesn't work, because old and crappy compilers get scared when they see that kind of thing.

Before anyone asks, yes, `<version>` is a must. I've had troubles with clang with `-std=c++20`.